src/plugins/mer/merqmakebuildconfiguration.cpp

```diff
@@ -301,10 +303,22 @@ void MerQmakeBuildConfiguration::updateExtraParserArguments()
     qs->setRecursive(true);
 }
 
+void MerQmakeBuildConfiguration::disableQmakeSystem()
+{
+    for (auto aspect : aspects()) {
+        if (aspect->settingsKey() == "RunSystemFunction") {
+            TriStateAspect* runSystemAspect = qobject_cast<TriStateAspect*>(aspect);
+            runSystemAspect->setSetting(TriState::Disabled);
+            runSystemAspect->setVisible(false);
+        }
+    }
```

Comment:
done

Reply:
ah, only now I realized we should use algorithms. This might work:

```
    auto *runSystemAspect =  qobject_cast<TriStateAspect *>(Utils::findOrDefault(aspects(),
            Utils::equal(&BaseAspect::settingsKey, "RunSystemFunction")));
```

Reply:
I don't really see how that's better than a simple for loop, but whatever

Reply:
Oh I see, there's the added benefit of getting these nice and understandable compilation errors: 
In file included from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/../../src/corelib/tools/qhashfunctions.h:48,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/qhashfunctions.h:1,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/../../src/corelib/tools/qlist.h:47,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/qlist.h:1,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/../../src/corelib/kernel/qobject.h:49,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/qobject.h:1,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/../../src/corelib/kernel/qcoreapplication.h:46,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/qcoreapplication.h:1,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/QCoreApplication:1,
                 from ../../../../sailfish-qtcreator/src/libs/utils/fileutils.h:32,
                 from ../../../../sailfish-qtcreator/src/libs/utils/aspects.h:28,
                 from ../../../../sailfish-qtcreator/src/plugins/projectexplorer/projectconfiguration.h:30,
                 from ../../../../sailfish-qtcreator/src/plugins/projectexplorer/buildconfiguration.h:29,
                 from ../../../../sailfish-qtcreator/src/plugins/qmakeprojectmanager/qmakebuildconfiguration.h:30,
                 from ../../../../sailfish-qtcreator/src/plugins/mer/merqmakebuildconfiguration.h:26,
                 from ../../../../sailfish-qtcreator/src/plugins/mer/merqmakebuildconfiguration.cpp:24:
/usr/include/c++/8/functional: In instantiation of ‘std::_Bind_result<_Result, _Functor(_Bound_args ...)>::__disable_if_void<_Res> std::_Bind_result<_Result, _Functor(_Bound_args ...)>::__call(std::tuple<_Args2 ...>&&, std::_Index_tuple<_Indexes ...>) [with _Res = bool; _Args = {Utils::BaseAspect* const&}; long unsigned int ..._Indexes = {0, 1}; _Result = bool; _Functor = std::equal_to<const char*>; _Bound_args = {const char*, std::_Bind<QString (Utils::BaseAspect::*(std::_Placeholder<1>))() const>}; std::_Bind_result<_Result, _Functor(_Bound_args ...)>::__disable_if_void<_Res> = bool]’:
/usr/include/c++/8/functional:651:24:   required from ‘std::_Bind_result<_Result, _Functor(_Bound_args ...)>::result_type std::_Bind_result<_Result, _Functor(_Bound_args ...)>::operator()(_Args&& ...) [with _Args = {Utils::BaseAspect* const&}; _Result = bool; _Functor = std::equal_to<const char*>; _Bound_args = {const char*, std::_Bind<QString (Utils::BaseAspect::*(std::_Placeholder<1>))() const>}; std::_Bind_result<_Result, _Functor(_Bound_args ...)>::result_type = bool]’
/usr/include/c++/8/bits/predefined_ops.h:283:11:   required from ‘bool __gnu_cxx::__ops::_Iter_pred<_Predicate>::operator()(_Iterator) [with _Iterator = QList<Utils::BaseAspect*>::const_iterator; _Predicate = std::_Bind_result<bool, std::equal_to<const char*>(const char*, std::_Bind<QString (Utils::BaseAspect::*(std::_Placeholder<1>))() const>)>]’
/usr/include/c++/8/bits/stl_algo.h:120:14:   required from ‘_RandomAccessIterator std::__find_if(_RandomAccessIterator, _RandomAccessIterator, _Predicate, std::random_access_iterator_tag) [with _RandomAccessIterator = QList<Utils::BaseAspect*>::const_iterator; _Predicate = __gnu_cxx::__ops::_Iter_pred<std::_Bind_result<bool, std::equal_to<const char*>(const char*, std::_Bind<QString (Utils::BaseAspect::*(std::_Placeholder<1>))() const>)> >]’
/usr/include/c++/8/bits/stl_algo.h:161:23:   required from ‘_Iterator std::__find_if(_Iterator, _Iterator, _Predicate) [with _Iterator = QList<Utils::BaseAspect*>::const_iterator; _Predicate = __gnu_cxx::__ops::_Iter_pred<std::_Bind_result<bool, std::equal_to<const char*>(const char*, std::_Bind<QString (Utils::BaseAspect::*(std::_Placeholder<1>))() const>)> >]’
/usr/include/c++/8/bits/stl_algo.h:3930:28:   required from ‘_IIter std::find_if(_IIter, _IIter, _Predicate) [with _IIter = QList<Utils::BaseAspect*>::const_iterator; _Predicate = std::_Bind_result<bool, std::equal_to<const char*>(const char*, std::_Bind<QString (Utils::BaseAspect::*(std::_Placeholder<1>))() const>)>]’
../../../../sailfish-qtcreator/src/libs/utils/algorithm.h:478:49:   required from ‘typename C::value_type Utils::findOr(const C&, typename C::value_type, F) [with C = Utils::BaseAspects; F = std::_Bind_result<bool, std::equal_to<const char*>(const char*, std::_Bind<QString (Utils::BaseAspect::*(std::_Placeholder<1>))() const>)>; typename C::value_type = Utils::BaseAspect*]’
../../../../sailfish-qtcreator/src/libs/utils/algorithm.h:505:18:   required from ‘std::enable_if_t<std::is_copy_assignable<typename C::value_type>::value, typename C::value_type> Utils::findOrDefault(const C&, F) [with C = Utils::BaseAspects; F = std::_Bind_result<bool, std::equal_to<const char*>(const char*, std::_Bind<QString (Utils::BaseAspect::*(std::_Placeholder<1>))() const>)>; std::enable_if_t<std::is_copy_assignable<typename C::value_type>::value, typename C::value_type> = Utils::BaseAspect*]’
../../../../sailfish-qtcreator/src/plugins/mer/merqmakebuildconfiguration.cpp:309:80:   required from here
/usr/include/c++/8/functional:556:24: error: no matching function for call to ‘__invoke(std::equal_to<const char*>&, const char*&, QString)’
    return std::__invoke(_M_f, _Mu<_Bound_args>()
           ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
         (std::get<_Indexes>(_M_bound_args), __args)...);
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/8/tuple:41,
                 from /usr/include/c++/8/functional:54,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/../../src/corelib/tools/qhashfunctions.h:48,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/qhashfunctions.h:1,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/../../src/corelib/tools/qlist.h:47,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/qlist.h:1,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/../../src/corelib/kernel/qobject.h:49,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/qobject.h:1,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/../../src/corelib/kernel/qcoreapplication.h:46,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/qcoreapplication.h:1,
                 from /home/vige/invariant/qt-everywhere-src-5.15.2/qtbase/include/QtCore/QCoreApplication:1,
                 from ../../../../sailfish-qtcreator/src/libs/utils/fileutils.h:32,
                 from ../../../../sailfish-qtcreator/src/libs/utils/aspects.h:28,
                 from ../../../../sailfish-qtcreator/src/plugins/projectexplorer/projectconfiguration.h:30,
                 from ../../../../sailfish-qtcreator/src/plugins/projectexplorer/buildconfiguration.h:29,
                 from ../../../../sailfish-qtcreator/src/plugins/qmakeprojectmanager/qmakebuildconfiguration.h:30,
                 from ../../../../sailfish-qtcreator/src/plugins/mer/merqmakebuildconfiguration.h:26,
                 from ../../../../sailfish-qtcreator/src/plugins/mer/merqmakebuildconfiguration.cpp:24:
/usr/include/c++/8/bits/invoke.h:89:5: note: candidate: ‘template<class _Callable, class ... _Args> constexpr typename std::__invoke_result<_Functor, _ArgTypes>::type std::__invoke(_Callable&&, _Args&& ...)’
     __invoke(_Callable&& __fn, _Args&&... __args)
     ^~~~~~~~
/usr/include/c++/8/bits/invoke.h:89:5: note:   template argument deduction/substitution failed:
/usr/include/c++/8/bits/invoke.h: In substitution of ‘template<class _Callable, class ... _Args> constexpr typename std::__invoke_result<_Functor, _ArgTypes>::type std::__invoke(_Callable&&, _Args&& ...) [with _Callable = std::equal_to<const char*>&; _Args = {const char*&, QString}]’:
/usr/include/c++/8/functional:556:24:   required from ‘std::_Bind_result<_Result, _Functor(_Bound_args ...)>::__disable_if_void<_Res> std::_Bind_result<_Result, _Functor(_Bound_args ...)>::__call(std::tuple<_Args2 ...>&&, std::_Index_tuple<_Indexes ...>) [with _Res = bool; _Args = {Utils::BaseAspect* const&}; long unsigned int ..._Indexes = {0, 1}; _Result = bool; _Functor = std::equal_to<const char*>; _Bound_args = {const char*, std::_Bind<QString (Utils::BaseAspect::*(std::_Placeholder<1>))() const>}; std::_Bind_result<_Result, _Functor(_Bound_args ...)>::__disable_if_void<_Res> = bool]’
/usr/include/c++/8/functional:651:24:   required from ‘std::_Bind_result<_Result, _Functor(_Bound_args ...)>::result_type std::_Bind_result<_Result, _Functor(_Bound_args ...)>::operator()(_Args&& ...) [with _Args = {Utils::BaseAspect* const&}; _Result = bool; _Functor = std::equal_to<const char*>; _Bound_args = {const char*, std::_Bind<QString (Utils::BaseAspect::*(std::_Placeholder<1>))() const>}; std::_Bind_result<_Result, _Functor(_Bound_args ...)>::result_type = bool]’
/usr/include/c++/8/bits/predefined_ops.h:283:11:   required from ‘bool __gnu_cxx::__ops::_Iter_pred<_Predicate>::operator()(_Iterator) [with _Iterator = QList<Utils::BaseAspect*>::const_iterator; _Predicate = std::_Bind_result<bool, std::equal_to<const char*>(const char*, std::_Bind<QString (Utils::BaseAspect::*(std::_Placeholder<1>))() const>)>]’
/usr/include/c++/8/bits/stl_algo.h:120:14:   required from ‘_RandomAccessIterator std::__find_if(_RandomAccessIterator, _RandomAccessIterator, _Predicate, std::random_access_iterator_tag) [with _RandomAccessIterator = QList<Utils::BaseAspect*>::const_iterator; _Predicate = __gnu_cxx::__ops::_Iter_pred<std::_Bind_result<bool, std::equal_to<const char*>(const char*, std::_Bind<QString (Utils::BaseAspect::*(std::_Placeholder<1>))() const>)> >]’
/usr/include/c++/8/bits/stl_algo.h:161:23:   required from ‘_Iterator std::__find_if(_Iterator, _Iterator, _Predicate) [with _Iterator = QList<Utils::BaseAspect*>::const_iterator; _Predicate = __gnu_cxx::__ops::_Iter_pred<std::_Bind_result<bool, std::equal_to<const char*>(const char*, std::_Bind<QString (Utils::BaseAspect::*(std::_Placeholder<1>))() const>)> >]’
/usr/include/c++/8/bits/stl_algo.h:3930:28:   required from ‘_IIter std::find_if(_IIter, _IIter, _Predicate) [with _IIter = QList<Utils::BaseAspect*>::const_iterator; _Predicate = std::_Bind_result<bool, std::equal_to<const char*>(const char*, std::_Bind<QString (Utils::BaseAspect::*(std::_Placeholder<1>))() const>)>]’
../../../../sailfish-qtcreator/src/libs/utils/algorithm.h:478:49:   required from ‘typename C::value_type Utils::findOr(const C&, typename C::value_type, F) [with C = Utils::BaseAspects; F = std::_Bind_result<bool, std::equal_to<const char*>(const char*, std::_Bind<QString (Utils::BaseAspect::*(std::_Placeholder<1>))() const>)>; typename C::value_type = Utils::BaseAspect*]’
../../../../sailfish-qtcreator/src/libs/utils/algorithm.h:505:18:   required from ‘std::enable_if_t<std::is_copy_assignable<typename C::value_type>::value, typename C::value_type> Utils::findOrDefault(const C&, F) [with C = Utils::BaseAspects; F = std::_Bind_result<bool, std::equal_to<const char*>(const char*, std::_Bind<QString (Utils::BaseAspect::*(std::_Placeholder<1>))() const>)>; std::enable_if_t<std::is_copy_assignable<typename C::value_type>::value, typename C::value_type> = Utils::BaseAspect*]’
../../../../sailfish-qtcreator/src/plugins/mer/merqmakebuildconfiguration.cpp:309:80:   required from here
/usr/include/c++/8/bits/invoke.h:89:5: error: no type named ‘type’ in ‘struct std::__invoke_result<std::equal_to<const char*>&, const char*&, QString>’

Reply:
The for loop works fine, I'm not going to spend hours trying to find out what's wrong with your algorithmic approach

Reply:
Don't get me wrong, it's not my approach, it's the trend (not only) in qtc upstream :) I guess in this case passing `QString("RunSystemFunction")` should help.

Reply:
Yup, it was indeed the missing QString there. Sorry, I didn't really mean to blame you for the qtc upstream style - even if you might think I did because I wrote "your algorithmic approach". I still don't like it though.

Reply:
Their intention is to "Say what should be done, rather than just how it should be done." https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#p3-express-intent . E.g. in our case one can read directly "find aspect where settings key equals to ..." instead of "iterate over all aspect ...". You know, written once, read many times. It's like goto vs for loop - both work well but the latter leads to a bit more readable code :) Pity the standard library isn't there yet to get the best out of it, but e.g. QtC's utils really do a nice job here.

Reply:
Might depend on the case, but I think the for loop was much more readable/easier to understand. QtC Utils are not bad, but I don't think they should be used for *everything*. But of course, if your only tool is a hammer, then all problems are nails... 

I wish they could do something to those compiler errors. Well, in many cases the errors from the stl which used to have similar explosions are much better these days, so I guess there's still hope.

Reply:
Yup, there is a hope https://en.wikipedia.org/wiki/Concepts_(C%2B%2B)

Reply:
I had a vague recollection of reading about concepts and how they can be used for improving the compiler error messages. I just couldn't pinpoint what it was exactly, i.e. I didn't remember that it was about concepts. Thanks for reminding me about that.

I don't really like the way they have been developing C++ in the last decade, i.e. adding every imaginable feature there, often with clumsy syntax. But there are gold nuggets there every now and then, e.g. lambdas. And I have a feeling concepts might be one of those. I don't even care if it's a clumsy syntax, as most of the time you don't have to care about that as you are not writing them - you just get the benefits when you use templates which use concepts :)